course/src/isometry.cc

```diff
@@ -0,0 +1,288 @@
+    return Vector3(x_ / v.x_, y_ / v.y_, z_ / v.z_);
+}
+
+Vector3& Vector3::operator = (const Vector3& v) {
+    x_ = v.x();
+    y_ = v.y();
+    z_ = v.z();
+    return *this;
+}
+
+Vector3& ekumen::math::Vector3::operator += (const Vector3& v) { 
+    x_ += v.x_;
+    y_ += v.y_;
+    z_ += v.z_;
+    return *this;
+}
+
+Vector3& ekumen::math::Vector3::operator -= (const Vector3& v) {
+    x_ -= v.x_;
+    y_ -= v.y_;
+    z_ -= v.z_;
+    return *this; 
+}
+
+Vector3& ekumen::math::Vector3::operator *= (const Vector3& v) {
+    x_ *= v.x_;
+    y_ *= v.y_;
+    z_ *= v.z_;
+    return *this; 
+}
+
+Vector3& ekumen::math::Vector3::operator /= (const Vector3& v) {
+    x_ /= v.x_;
+    y_ /= v.y_;
+    z_ /= v.z_;
+    return *this; 
+}
+
+bool ekumen::math::Vector3::operator == (const Vector3& v) const {
+    return (x_ == v.x_) && (y_ == v.y_) && (z_ == v.z_);
+}
+
+bool ekumen::math::Vector3::operator != (const Vector3& v) const {
+    return !(*this == v);
+}
+
+bool ekumen::math::Vector3::operator == (const std::initializer_list<double>& list) const {
+    if(list.size() != 3) {
+        throw "Invalid initializer list size";
+    }
+
+    return (*this) == Vector3(list);
+}
+
+bool ekumen::math::Vector3::operator != (const std::initializer_list<double>& list) const {
+    return !(*this == list);
+}
+
+// Computations
+double Vector3::dot(const Vector3& v) const {
+    return (x_ * v.x_) + (y_ * v.y_) + (z_ * v.z_);
+}
+
+double Vector3::norm() const {
+    return sqrt( (x_ * x_) + (y_ * y_) + (z_* z_) );
+}
+
+Vector3 Vector3::cross(const Vector3& v) const {
+    return Vector3(
+    y_ * v.z_ - z_ * v.y_, 
+    z_ * v.x_ - x_ * v.z_,
+    x_ * v.y_ - y_ * v.x_);
+}
+
+// Class constants
+const Vector3 Vector3::kUnitX = Vector3(1,0,0);
+const Vector3 Vector3::kUnitY = Vector3(0,1,0);
+const Vector3 Vector3::kUnitZ = Vector3(0,0,1);
+const Vector3 Vector3::kZero = Vector3(0,0,0);
+
+
+
+
+
+
+// Matrix3 class methods
+
+// Constructors
+Matrix3::Matrix3(const std::initializer_list<double>& list) {
+    if(list.size() != 9) {
+        throw "Invalid initializer list size";
+    }
+    std::initializer_list<double>::iterator it = list.begin();
+    r1_ = Vector3(std::initializer_list<double>{*it++, *it++, *it++});
+    r2_ = Vector3(std::initializer_list<double>{*it++, *it++, *it++});
+    r3_ = Vector3(std::initializer_list<double>{*it++, *it++, *it});
+}
+
+Matrix3::Matrix3(const std::initializer_list<double>& l1,
+                 const std::initializer_list<double>& l2,
+                 const std::initializer_list<double>& l3) {
+    r1_ = Vector3(l1);
+    r2_ = Vector3(l2);
+    r3_ = Vector3(l3);
+}
+
+// Operators
+Vector3& Matrix3::operator [] (const int index) {
+    return this->row(index);
+}
+
+const Vector3& Matrix3::operator [] (const int index) const {
+    return this->row(index);
+}
+
+Matrix3 Matrix3::operator + (const Matrix3& m) const {
+    return Matrix3(r1_ + m.r1_, r2_ + m.r2_, r3_ + m.r3_);
```

Comment:
Use `+=` to implement this.

Reply:
Mhmm... But this method is const. If I use += it wouldn't be const.

Reply:
But you can create a local variable that is a copy of `*this` and change that.
```
const Matrix3 &Matrix3::operator + (const Matrix3& m) const {
  Matrix lhs(*this);  
  lhs += m;
  return lhs;
}
```